vendor: use go mod

errordeveloper opened the pull request.

errordeveloper commented on vendor.mod:
```diff
@@ -0,0 +1,198 @@
+	google.golang.org/genproto => google.golang.org/genproto v0.0.0-20200227132054-3f1135a288c9
+	google.golang.org/grpc => google.golang.org/grpc v1.27.1
+	gopkg.in/fsnotify.v1 => gopkg.in/fsnotify.v1 v1.4.7
+        archive/tar => ./internal/archive/tar
```
So this actually isn't supported (see https://github.com/golang/go/issues/35283). We need to find a better way to do this, perhaps for now there could be `cp -a internal/archive/tar` inside of `go mod vendor` wrapper.

thaJeztah replied:
If we keep the files inside `vendor/`, does `go mod vendor` delete them, or does it leave them alone?

also wondering if a symlink would work (although Windows may not be happy with that, I recall there were some issues with Windows and symlinks, but don't recall the details)

errordeveloper replied:
> If we keep the files inside `vendor/`, does `go mod vendor` delete them, or does it leave them alone?

It deletes them.

> also wondering if a symlink would work (although Windows may not be happy with that, I recall there were some issues with Windows and symlinks, but don't recall the details)

The issue with symlinks on Windows is pretty major IMO, a Windows user basically cannot clone a repo with symlinks.

AkihiroSuda replied:
Maybe I asked before, but why can't we move `vendor/archive/tar` to an external module such as `github.com/moby/sys/archive/tar` and just apply `s|archive/tar|github.com/moby/sys/archive/tar|g` to the moby/moby source?

errordeveloper replied:
I think it might be worse trying as a follow-up, but here the goal is to make the least number of changes.

thaJeztah replied:
@AkihiroSuda In that case, we could as well put it in `internal/` in this repository. But I think the reason we didn't "fork" it (but override) is that it may be used indirectly(?). Are we sure that the archive/tar code is never called indirectly (and we would reintroduce the vulnerability?)

thaJeztah commented:
Also opened https://github.com/docker/docker-ce-packaging/pull/611 to do a quick check if the packaging scripts don't blow up 👍

crazy-max commented on hack/vendor.sh:
```diff
@@ -1,28 +1,37 @@
+go 1.17
+EOF
+
+cat > hack/make/.resources-windows/go.mod << EOF
+module github.com/docker/docker/autogen/winresources/dockerd
```
For this I can open an issue to implement goversioninfo here like it has be done in https://github.com/docker/cli/pull/3310 so you would not need to add this hack. WDYT @thaJeztah?

thaJeztah replied:
Yes, would be great if we could used the same tool.

I'm ok with a temporary hack like this (if it works in the meantime)

errordeveloper replied:
I think it'd be good as a follow-up, but doesn't seem like a blocker...

errordeveloper commented:
This should be good as far as I'm concerned, but let's see what CI might bring...

errordeveloper commented on hack/vendor.sh:
```diff
@@ -1,28 +1,37 @@
+go 1.17
+EOF
 
 if [ $# -eq 0 ] || [ "$1" = "archive/tar" ]; then
```
@thaJeztah I think this should do the trick!

errordeveloper replied:
Actually, I've misunderstood this initially, looks like `internal/archive/tar` is not really needed, since download & patching is done each time this script runs... I was assuming it's options (clearly didn't read this script fully).

thaJeztah commented on hack/vendor.sh:
```diff
@@ -1,28 +1,37 @@
 
 if [ $# -eq 0 ] || [ "$1" != "archive/tar" ]; then
-	vndr -whitelist='^archive[/\\]tar' "$@"
+	go mod vendor -modfile vendor.mod
```
Not 100% sure; do we need a `go mod tidy` before `go mod vendor` ?

Looks like we need to set `GO111MODULES=auto` here. When I try to run it inside the "dev" container (`make BIND_DIR=. shell`), it complains that modules are disabled;

```
hack/vendor.sh
update vendored copy of archive/tar
downloading: https://golang.org/dl/go1.17.4.src.tar.gz
patching file stat_unix.go
Hunk #1 succeeded at 9 (offset 1 line).
Hunk #2 succeeded at 17 (offset 1 line).
Hunk #3 succeeded at 25 (offset 1 line).
go: modules disabled by GO111MODULE=off; see 'go help modules'
```

errordeveloper replied:
I think `go mod tidy` should be called explicitly, it has too much of a side-effect :)

errordeveloper replied:
> Looks like we need to set `GO111MODULES=auto` here. When I try to run it inside the "dev" container (`make BIND_DIR=. shell`), it complains that modules are disabled;

fixed 👍

errordeveloper replied:
This seems to happen if you run it without `-modfile vendor.mod`, but with that flag these errors went away when I check last time.

errordeveloper replied:
Per @crazy-max's suggestions I've just added `go mod tidy` to `hack/validate/vendor`, hope that addresses your concern.

thaJeztah commented on hack/vendor.sh:
```diff
@@ -1,28 +1,37 @@
 fi
+
+rm -rf vendor/archive/tar
+cp -a internal/archive/tar vendor/archive/tar
```
Looks like it's failing here;

```
cp: cannot create directory 'vendor/archive/tar': No such file or directory
```

```suggestion
mkdir -p
cp -a internal/archive/tar vendor/archive/tar
```

That said, I'm a bit confused, because in this branch, `vendor/archive/tar` doesn't exist (only `internal/archive/tar`), and we don't reference `internal/archive/tar`, or have the `replace` rule anymore?

Should the script be rewritten to use a temp-dir for the git clone, and then move the files? (and don't have the rename `vendor/archive/... -> internal/archive/...` in this branch?

thaJeztah reviewed (changes requested):
Thanks! Did some quick testing, and looks like there may be some issues (see comments)

errordeveloper commented on hack/validate/vendor:
```diff
@@ -39,16 +43,15 @@ validate_vendor_diff(){
 	done
 }
 
+go_mod_tidy
```
@thaJeztah @crazy-max  PTAL 👁️

errordeveloper replied:
@thaJeztah @crazy-max  PTAL 👁️

errordeveloper commented on hack/validate/vendor:
```diff
@@ -3,10 +3,14 @@
 source "${SCRIPTDIR}/.validate"
 
+go_mod_tidy(){
+	GO111MODULE=auto go mod vendor -modfile 'vendor.mod'
```
```suggestion
	GO111MODULE=auto go mod tidy -modfile 'vendor.mod'
```

errordeveloper commented:
@thaJeztah do you know of `hack/validate/vendor` runs in Jenkins without `hack/vendor.sh` running prior to it? I was expecting `hack/vendor.sh` to run first, but based on what I see from the last failed build, that is not the case...

errordeveloper commented:
> @thaJeztah do you know of `hack/validate/vendor` runs in Jenkins without `hack/vendor.sh` running prior to it?

I can see it is that way, yes. Fortunately Jenkinsfile was easy enough to read :)

tianon commented on hack/vendor.sh:
```diff
@@ -1,28 +1,30 @@
-	patch --strip=4 --directory=vendor/archive/tar --input="$PWD/patches/0001-archive-tar-do-not-populate-user-group-names.patch"
+		| tar --extract --gzip --directory=internal/archive/tar --strip-components=4 go/src/archive/tar
+	patch --strip=4 --directory=internal/archive/tar --input="$PWD/patches/0001-archive-tar-do-not-populate-user-group-names.patch"
+        rm -f internal/archive/tar/*_test.go
```
Whitespace on this line isn't quite right -- should be a single tab:

```suggestion
	rm -f internal/archive/tar/*_test.go
```

errordeveloper replied:
Ah, of course...

tianon commented on hack/vendor.sh:
```diff
@@ -1,28 +1,30 @@
 
 if [ $# -eq 0 ] || [ "$1" != "archive/tar" ]; then
-	vndr -whitelist='^archive[/\\]tar' "$@"
+        GO111MODULE=auto go mod vendor -modfile vendor.mod
```
```suggestion
	GO111MODULE=auto go mod vendor -modfile vendor.mod
```

tianon commented on hack/validate/vendor:
```diff
@@ -3,10 +3,15 @@
 source "${SCRIPTDIR}/.validate"
 
+go_mod_tidy(){
+	${SCRIPTDIR}/../go-mod-prepare.sh
```
Looks like this is missing some quotes, and the file itself doesn't appear to be included in this PR?

errordeveloper commented:
@crazy-max @thaJeztah please see last two commits, looks an okay solution to me - what do you think? adding `go.mod` to libkv happens to break consul, so I'm leaving that out for now in favour or updating just the zookeeper client.

errordeveloper committed:
vendor: replace vndr with `go mod vendor`

- use `vendor.mod` instead of `go.mod` to avoid issues to do with
  use of CalVer, not SemVer
- ensure most of the dependency versions do not change
  - only zookeeper client has to change (via docker/libkv#218) as
    previously used version is no longer maintained and has missing
    dependencies

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
vendor: regenerate

- all changes here are attributed to difference in behaviour between,
  namely:
  - resolution of secondary test dependencies
  - prunning of non-Go files

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

crazy-max commented on hack/vendor.sh:
```diff
@@ -1,28 +1,28 @@
+        curl -fsSL "https://golang.org/dl/go${GO_VERSION%.0}.src.tar.gz" \
+                | tar --extract --gzip --directory=vendor/archive/tar --strip-components=4 go/src/archive/tar
+        patch --strip=4 --directory=vendor/archive/tar --input="$PWD/patches/0001-archive-tar-do-not-populate-user-group-names.patch"
+        rm -f vendor/archive/tar/*_test.go
```
Small nit and maybe not necessary but we should use the `find` cmd in case of subpkg in the future:

```console
find ./vendor/archive/tar -name "*_test.go" -type f -exec rm -f {} \;
```

thaJeztah commented on hack/vendor.sh:
```diff
@@ -1,28 +1,29 @@
-		| tar --extract --gzip --directory=vendor/archive/tar --strip-components=4 go/src/archive/tar
-	patch --strip=4 --directory=vendor/archive/tar --input="$PWD/patches/0001-archive-tar-do-not-populate-user-group-names.patch"
-fi
+        echo "update vendored copy of archive/tar"
```
Looks like this changed from spaces to tabs 🙈

errordeveloper replied:
I only ever do the other way around! ;)

errordeveloper replied:
Will fix...

thaJeztah commented on hack/vendor.sh:
```diff
@@ -1,28 +1,29 @@
+        curl -fsSL "https://golang.org/dl/go${GO_VERSION%.0}.src.tar.gz" \
+                | tar --extract --gzip --directory=vendor/archive/tar --strip-components=4 go/src/archive/tar
+        patch --strip=4 --directory=vendor/archive/tar --input="$PWD/patches/0001-archive-tar-do-not-populate-user-group-names.patch"
+        find vendor/archive/tar -name "*_test.go" -type f -exec rm -f {} \;
```
I noticed this during the meeting; looks like before this PR, the `_test.go` file were kept (admitted; they won't be used); we could decide to just leave them be (and reduce the changes we make to it);  (guess I'm fine either way)

thaJeztah commented on hack/go-mod-prepare.sh:
```diff
@@ -0,0 +1,15 @@
+ROOTDIR="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
+
+cat > "${ROOTDIR}/go.mod" << EOF
+module github.com/moby/moby
```
I realize this is just to satisfy go mod; but just noticed this is using moby/moby; should it be docker/docker? Or no difference whatsoever?

errordeveloper replied:
I don't think it matters, it all works fine. When I started working on this I didn't realise all import paths were still docker/docker ;)

errordeveloper replied:
I just noticed that the other `go.mod` file in this script is written with docker/docker, so I've made these two consistent  :)

errordeveloper commented:
So docker/libkv#218 was merged, and I deforked it here. I also addressed other cosmetic concerns. I think this is 100% ready to go now!

crazy-max commented:
> So [docker/libkv#218](https://github.com/docker/libkv/pull/218) was merged

You mean https://github.com/docker/libkv/pull/219 right?

errordeveloper commented:
> You mean https://github.com/docker/libkv/pull/219 right?

Yes!

errordeveloper commented:
Only Widows tests seem to fail in CI, and that is quite clearly a file-access race which is out to be a flake:
```
[2021-12-20T11:32:48.134Z] === Failed
[2021-12-20T11:32:48.134Z] === FAIL: github.com/docker/docker/daemon/logger/jsonfilelog TestJSONFileLoggerWithOpts (0.02s)
[2021-12-20T11:32:48.134Z]     jsonfilelog_test.go:187: open C:\Users\ContainerAdministrator\AppData\Local\Temp\docker-logger-112550856\container.log.1: The process cannot access the file because it is being used by another process.
[2021-12-20T11:32:48.134Z] 
[2021-12-20T11:32:48.134Z] DONE 2099 tests, 70 skipped, 1 failure in 298.336s
```

thaJeztah commented:
<img width="201" alt="Screenshot 2021-12-20 at 16 32 01" src="https://user-images.githubusercontent.com/1804568/146793257-065bea23-4649-42f2-8b16-3b338e711ff4.png">

Had a look at the large diff in vendor. Looks like many of those lines are because `go mod vendor` also pulls in a lot of random files (`.sh`, any `.md` file, any `.yaml` file, as well as Makefiles etc etc). Not much we can do about that (perhaps we should open a ticket in Golang to suggest skipping some of those, as some are really useless as part of `vendor`).

There's also some dependencies that we previously were able to avoid (by only consuming certain packages), but are now pulled in because `go.mod` mentions them and/or because some repositories are using a `tools.go` (or equivalent) to have an easy way to download test dependencies.


This one we _may_ be able to control (if we would be able to separate the CLI parts of buildkit into a separate module);

```bash
go mod graph -modfile=vendor.mod | grep ' github.com/BurntSushi/toml'
```

<details>
```
github.com/docker/docker github.com/BurntSushi/toml@v0.3.1
github.com/moby/buildkit@v0.8.2-0.20210615162540-9f254e18360a github.com/BurntSushi/toml@v0.3.1
github.com/urfave/cli@v1.22.2 github.com/BurntSushi/toml@v0.3.1
github.com/urfave/cli@v1.22.1 github.com/BurntSushi/toml@v0.3.1
github.com/urfave/cli@v1.22.4 github.com/BurntSushi/toml@v0.3.1
```
</details>

There's also quite some dependencies coming in through `cloud.google.com/go` (Cloud provider SDK's are the worst, as they tend to be somewhat "mono-repo", and have a large dependency tree.

It looks like the `cloud.google.com/go` is used by various dependencies, but many of them _only_ for the `cloud.google.com/go/compute/metadata` package, which itself has no external dependencies, but (unlike various other parts in that repository) does not have its own `go.mod` :disappointed:  https://github.com/googleapis/google-cloud-go/tree/main/compute/metadata. Perhaps something we should look at, and see if they would be open to make that package a separate module (although we may need to test before doing soo to see if that works, and it would still take time before all indirect dependencies migrated)

```bash
go mod graph -modfile=vendor.mod | grep ' github.com/jstemmer/go-junit-report'
```

<details>
```
github.com/docker/docker github.com/jstemmer/go-junit-report@v0.9.1
cloud.google.com/go@v0.81.0 github.com/jstemmer/go-junit-report@v0.0.0-20190106144839-af01ea7f8024
cloud.google.com/go@v0.44.1 github.com/jstemmer/go-junit-report@v0.0.0-20190106144839-af01ea7f8024
cloud.google.com/go@v0.43.0 github.com/jstemmer/go-junit-report@v0.0.0-20190106144839-af01ea7f8024
cloud.google.com/go@v0.34.0 github.com/jstemmer/go-junit-report@v0.0.0-20190106144839-af01ea7f8024
cloud.google.com/go@v0.38.0 github.com/jstemmer/go-junit-report@v0.0.0-20190106144839-af01ea7f8024
cloud.google.com/go@v0.54.0 github.com/jstemmer/go-junit-report@v0.0.0-20190106144839-af01ea7f8024
cloud.google.com/go@v0.46.3 github.com/jstemmer/go-junit-report@v0.0.0-20190106144839-af01ea7f8024
```
</details>

```bash
go mod graph -modfile=vendor.mod | grep ' golang.org/x/exp'
```

<details>
```
github.com/docker/docker golang.org/x/exp@v0.0.0-20191030013958-a1ab85dbe136
cloud.google.com/go@v0.81.0 golang.org/x/exp@v0.0.0-20190510132918-efd6b22b2522
cloud.google.com/go@v0.44.1 golang.org/x/exp@v0.0.0-20190510132918-efd6b22b2522
cloud.google.com/go@v0.43.0 golang.org/x/exp@v0.0.0-20190510132918-efd6b22b2522
golang.org/x/mobile@v0.0.0-20190719004257-d2bd2a29d028 golang.org/x/exp@v0.0.0-20190306152737-a1d7652674e8
cloud.google.com/go@v0.34.0 golang.org/x/exp@v0.0.0-20190510132918-efd6b22b2522
cloud.google.com/go@v0.38.0 golang.org/x/exp@v0.0.0-20190510132918-efd6b22b2522
cloud.google.com/go@v0.54.0 golang.org/x/exp@v0.0.0-20190510132918-efd6b22b2522
cloud.google.com/go/firestore@v1.1.0 golang.org/x/exp@v0.0.0-20191030013958-a1ab85dbe136
cloud.google.com/go@v0.46.3 golang.org/x/exp@v0.0.0-20190510132918-efd6b22b2522
```
</details>

```bash
go mod graph -modfile=vendor.mod | grep ' golang.org/x/lint'
```

<details>
```
github.com/docker/docker golang.org/x/lint@v0.0.0-20201208152925-83fdc39ff7b5
cloud.google.com/go@v0.81.0 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
google.golang.org/api@v0.46.0 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
google.golang.org/genproto@v0.0.0-20211208223120-3a66f561d7aa golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
google.golang.org/api@v0.8.0 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
google.golang.org/genproto@v0.0.0-20190801165951-fa694d86fc64 golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
cloud.google.com/go@v0.44.1 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
cloud.google.com/go@v0.43.0 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
google.golang.org/api@v0.7.0 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
google.golang.org/genproto@v0.0.0-20190708153700-3bdd9d9f5532 golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
google.golang.org/genproto@v0.0.0-20200224152610-e50cd9704f63 golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
google.golang.org/genproto@v0.0.0-20190819201941-24fa4b261c55 golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
google.golang.org/genproto@v0.0.0-20201110150050-8816d57aaa9a golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
google.golang.org/genproto@v0.0.0-20190425155659-357c62f0e4bb golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
cloud.google.com/go@v0.34.0 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
cloud.google.com/go@v0.38.0 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
google.golang.org/genproto@v0.0.0-20190502173448-54afdca5d873 golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
go.etcd.io/etcd@v0.5.0-alpha.5.0.20200910180754-dd1b699fc489 golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
cloud.google.com/go@v0.54.0 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
google.golang.org/genproto@v0.0.0-20200526211855-cb27e3aa2013 golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
google.golang.org/genproto@v0.0.0-20200117163144-32f20d992d24 golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
cloud.google.com/go/firestore@v1.1.0 golang.org/x/lint@v0.0.0-20190930215403-16217165b5de
google.golang.org/api@v0.13.0 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
google.golang.org/genproto@v0.0.0-20180817151627-c66870c02cf8 golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
cloud.google.com/go@v0.46.3 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
google.golang.org/genproto@v0.0.0-20191108220845-16a3f7862a1a golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
google.golang.org/api@v0.9.0 golang.org/x/lint@v0.0.0-20190409202823-959b441ac422
google.golang.org/genproto@v0.0.0-20190911173649-1774047e7e51 golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3
```
</details>

```bash
go mod graph -modfile=vendor.mod | grep ' golang.org/x/mod'
```

<details>
```
github.com/docker/docker golang.org/x/mod@v0.4.1
golang.org/x/exp@v0.0.0-20191030013958-a1ab85dbe136 golang.org/x/mod@v0.1.0
golang.org/x/tools@v0.1.0 golang.org/x/mod@v0.3.0
golang.org/x/tools@v0.0.0-20210106214847-113979e3529a golang.org/x/mod@v0.3.0
golang.org/x/tools@v0.0.0-20191216052735-49a3e744a425 golang.org/x/mod@v0.1.1-0.20191105210325-c90efee705ee
golang.org/x/tools@v0.0.0-20200130002326-2f3ba24bd6e7 golang.org/x/mod@v0.1.1-0.20191105210325-c90efee705ee
golang.org/x/tools@v0.0.0-20200619180055-7c47624df98f golang.org/x/mod@v0.2.0
```
</details>

```bash
go mod graph -modfile=vendor.mod | grep ' golang.org/x/tools'
```

<details>
```
github.com/docker/docker golang.org/x/tools@v0.1.0
cloud.google.com/go@v0.81.0 golang.org/x/tools@v0.0.0-20190628153133-6cdbf07be9d0
github.com/gogo/protobuf@v1.3.2 golang.org/x/tools@v0.0.0-20210106214847-113979e3529a
github.com/miekg/dns@v1.1.27 golang.org/x/tools@v0.0.0-20191216052735-49a3e744a425
golang.org/x/exp@v0.0.0-20191030013958-a1ab85dbe136 golang.org/x/tools@v0.0.0-20191012152004-8de300cfc20a
golang.org/x/lint@v0.0.0-20201208152925-83fdc39ff7b5 golang.org/x/tools@v0.0.0-20200130002326-2f3ba24bd6e7
golang.org/x/mod@v0.4.1 golang.org/x/tools@v0.0.0-20191119224855-298f0cb1881e
golang.org/x/text@v0.3.6 golang.org/x/tools@v0.0.0-20180917221912-90fa682c2a6e
google.golang.org/api@v0.46.0 golang.org/x/tools@v0.0.0-20190506145303-2d16b83fe98c
google.golang.org/genproto@v0.0.0-20211208223120-3a66f561d7aa golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
gotest.tools/v3@v3.0.3 golang.org/x/tools@v0.0.0-20190624222133-a101b041ded4
github.com/golang/mock@v1.3.1 golang.org/x/tools@v0.0.0-20190425150028-36563e24a262
golang.org/x/exp@v0.0.0-20190510132918-efd6b22b2522 golang.org/x/tools@v0.0.0-20190312151545-0bb0c0a6e846
golang.org/x/lint@v0.0.0-20190409202823-959b441ac422 golang.org/x/tools@v0.0.0-20190311212946-11955173bddd
golang.org/x/text@v0.3.2 golang.org/x/tools@v0.0.0-20180917221912-90fa682c2a6e
google.golang.org/api@v0.8.0 golang.org/x/tools@v0.0.0-20190506145303-2d16b83fe98c
google.golang.org/genproto@v0.0.0-20190801165951-fa694d86fc64 golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
cloud.google.com/go@v0.44.1 golang.org/x/tools@v0.0.0-20190628153133-6cdbf07be9d0
cloud.google.com/go@v0.43.0 golang.org/x/tools@v0.0.0-20190628153133-6cdbf07be9d0
google.golang.org/api@v0.7.0 golang.org/x/tools@v0.0.0-20190506145303-2d16b83fe98c
google.golang.org/genproto@v0.0.0-20190708153700-3bdd9d9f5532 golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
google.golang.org/genproto@v0.0.0-20200224152610-e50cd9704f63 golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
github.com/kisielk/errcheck@v1.5.0 golang.org/x/tools@v0.0.0-20200619180055-7c47624df98f
google.golang.org/genproto@v0.0.0-20190819201941-24fa4b261c55 golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
github.com/miekg/dns@v1.1.26 golang.org/x/tools@v0.0.0-20190907020128-2ca718005c18
google.golang.org/genproto@v0.0.0-20201110150050-8816d57aaa9a golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
gotest.tools/v3@v3.0.2 golang.org/x/tools@v0.0.0-20190624222133-a101b041ded4
google.golang.org/genproto@v0.0.0-20190425155659-357c62f0e4bb golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
golang.org/x/text@v0.3.3 golang.org/x/tools@v0.0.0-20180917221912-90fa682c2a6e
cloud.google.com/go@v0.34.0 golang.org/x/tools@v0.0.0-20190628153133-6cdbf07be9d0
golang.org/x/mod@v0.3.0 golang.org/x/tools@v0.0.0-20191119224855-298f0cb1881e
cloud.google.com/go@v0.38.0 golang.org/x/tools@v0.0.0-20190628153133-6cdbf07be9d0
google.golang.org/genproto@v0.0.0-20190502173448-54afdca5d873 golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
golang.org/x/lint@v0.0.0-20190313153728-d0100b6bd8b3 golang.org/x/tools@v0.0.0-20190311212946-11955173bddd
google.golang.org/appengine@v1.6.1 golang.org/x/tools@v0.0.0-20190606124116-d0a3d012864b
golang.org/x/text@v0.3.4 golang.org/x/tools@v0.0.0-20180917221912-90fa682c2a6e
github.com/go-openapi/spec@v0.19.3 golang.org/x/tools@v0.0.0-20190614205625-5aca471b1d59
go.etcd.io/etcd@v0.5.0-alpha.5.0.20200910180754-dd1b699fc489 golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
cloud.google.com/go@v0.54.0 golang.org/x/tools@v0.0.0-20190628153133-6cdbf07be9d0
github.com/kisielk/errcheck@v1.2.0 golang.org/x/tools@v0.0.0-20181030221726-6c7e314b6563
github.com/kisielk/errcheck@v1.1.0 golang.org/x/tools@v0.0.0-20180221164845-07fd8470d635
github.com/smartystreets/goconvey@v1.6.4 golang.org/x/tools@v0.0.0-20190328211700-ab21143f2384
google.golang.org/genproto@v0.0.0-20200526211855-cb27e3aa2013 golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
golang.org/x/text@v0.3.0 golang.org/x/tools@v0.0.0-20180917221912-90fa682c2a6e
golang.org/x/mod@v0.2.0 golang.org/x/tools@v0.0.0-20191119224855-298f0cb1881e
google.golang.org/genproto@v0.0.0-20200117163144-32f20d992d24 golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
cloud.google.com/go/firestore@v1.1.0 golang.org/x/tools@v0.0.0-20191112195655-aa38f8e97acc
google.golang.org/api@v0.13.0 golang.org/x/tools@v0.0.0-20190506145303-2d16b83fe98c
google.golang.org/genproto@v0.0.0-20180817151627-c66870c02cf8 golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
cloud.google.com/go@v0.46.3 golang.org/x/tools@v0.0.0-20190628153133-6cdbf07be9d0
golang.org/x/lint@v0.0.0-20190930215403-16217165b5de golang.org/x/tools@v0.0.0-20190311212946-11955173bddd
google.golang.org/genproto@v0.0.0-20191108220845-16a3f7862a1a golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
google.golang.org/api@v0.9.0 golang.org/x/tools@v0.0.0-20190506145303-2d16b83fe98c
google.golang.org/genproto@v0.0.0-20190911173649-1774047e7e51 golang.org/x/tools@v0.0.0-20190524140312-2c0ae7006135
```
</details>

```bash
go mod graph -modfile=vendor.mod | grep ' golang.org/x/xerrors'
```

<details>
```
github.com/docker/docker golang.org/x/xerrors@v0.0.0-20200804184101-5ec99f83aff1
golang.org/x/mod@v0.4.1 golang.org/x/xerrors@v0.0.0-20191011141410-1b5146add898
golang.org/x/tools@v0.1.0 golang.org/x/xerrors@v0.0.0-20200804184101-5ec99f83aff1
golang.org/x/tools@v0.0.0-20210106214847-113979e3529a golang.org/x/xerrors@v0.0.0-20200804184101-5ec99f83aff1
golang.org/x/tools@v0.0.0-20191216052735-49a3e744a425 golang.org/x/xerrors@v0.0.0-20191011141410-1b5146add898
golang.org/x/tools@v0.0.0-20191012152004-8de300cfc20a golang.org/x/xerrors@v0.0.0-20190717185122-a985d3407aa7
golang.org/x/tools@v0.0.0-20200130002326-2f3ba24bd6e7 golang.org/x/xerrors@v0.0.0-20191011141410-1b5146add898
golang.org/x/tools@v0.0.0-20191119224855-298f0cb1881e golang.org/x/xerrors@v0.0.0-20190717185122-a985d3407aa7
golang.org/x/mod@v0.3.0 golang.org/x/xerrors@v0.0.0-20191011141410-1b5146add898
golang.org/x/tools@v0.0.0-20200619180055-7c47624df98f golang.org/x/xerrors@v0.0.0-20191204190536-9bdfabe68543
golang.org/x/tools@v0.0.0-20190907020128-2ca718005c18 golang.org/x/xerrors@v0.0.0-20190717185122-a985d3407aa7
golang.org/x/mod@v0.1.1-0.20191105210325-c90efee705ee golang.org/x/xerrors@v0.0.0-20191011141410-1b5146add898
golang.org/x/mod@v0.2.0 golang.org/x/xerrors@v0.0.0-20191011141410-1b5146add898
golang.org/x/tools@v0.0.0-20191112195655-aa38f8e97acc golang.org/x/xerrors@v0.0.0-20190717185122-a985d3407aa7
```
</details>


```bash
go mod graph -modfile=vendor.mod | grep ' google.golang.org/appengine'
```

<details>
```
github.com/docker/docker google.golang.org/appengine@v1.6.7
golang.org/x/oauth2@v0.0.0-20210514164344-f6687ab2804c google.golang.org/appengine@v1.4.0
google.golang.org/api@v0.46.0 google.golang.org/appengine@v1.5.0
cloud.google.com/go/datastore@v1.0.0 google.golang.org/appengine@v1.6.1
golang.org/x/oauth2@v0.0.0-20190604053449-0f29369cfe45 google.golang.org/appengine@v1.4.0
google.golang.org/api@v0.8.0 google.golang.org/appengine@v1.5.0
google.golang.org/api@v0.7.0 google.golang.org/appengine@v1.5.0
golang.org/x/oauth2@v0.0.0-20180821212333-d2e6202438be google.golang.org/appengine@v1.4.0
golang.org/x/oauth2@v0.0.0-20200107190931-bf48bf16ab8d google.golang.org/appengine@v1.4.0
google.golang.org/api@v0.13.0 google.golang.org/appengine@v1.5.0
google.golang.org/api@v0.9.0 google.golang.org/appengine@v1.5.0
```
</details>

```bash
go mod graph -modfile=vendor.mod | grep ' honnef.co/go/tools'
```

<details>
```
github.com/docker/docker honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
cloud.google.com/go@v0.81.0 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
google.golang.org/api@v0.46.0 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
google.golang.org/genproto@v0.0.0-20211208223120-3a66f561d7aa honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
google.golang.org/api@v0.8.0 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
google.golang.org/genproto@v0.0.0-20190801165951-fa694d86fc64 honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
cloud.google.com/go@v0.44.1 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
cloud.google.com/go@v0.43.0 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
google.golang.org/api@v0.7.0 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
google.golang.org/genproto@v0.0.0-20190708153700-3bdd9d9f5532 honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
google.golang.org/genproto@v0.0.0-20200224152610-e50cd9704f63 honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
google.golang.org/genproto@v0.0.0-20190819201941-24fa4b261c55 honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
google.golang.org/genproto@v0.0.0-20201110150050-8816d57aaa9a honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
google.golang.org/genproto@v0.0.0-20190425155659-357c62f0e4bb honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
cloud.google.com/go@v0.34.0 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
cloud.google.com/go@v0.38.0 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
google.golang.org/genproto@v0.0.0-20190502173448-54afdca5d873 honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
go.etcd.io/etcd@v0.5.0-alpha.5.0.20200910180754-dd1b699fc489 honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
cloud.google.com/go@v0.54.0 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
google.golang.org/genproto@v0.0.0-20200526211855-cb27e3aa2013 honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
google.golang.org/genproto@v0.0.0-20200117163144-32f20d992d24 honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
google.golang.org/api@v0.13.0 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
google.golang.org/genproto@v0.0.0-20180817151627-c66870c02cf8 honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
cloud.google.com/go@v0.46.3 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
google.golang.org/genproto@v0.0.0-20191108220845-16a3f7862a1a honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
google.golang.org/api@v0.9.0 honnef.co/go/tools@v0.0.0-20190418001031-e561f6794a2a
google.golang.org/genproto@v0.0.0-20190911173649-1774047e7e51 honnef.co/go/tools@v0.0.0-20190523083050-ea95bdfd59fc
```
</details>

thaJeztah commented:
I noticed that the temporary `go.mod` files that we generate are not cleaned up after `vendor.sh` completes. They are added to `.gitignore`, so not "critical", but perhaps (as a follow up?) we should clean those up, so that running `vendor.sh` leaves a clean state afterwards.

errordeveloper commented:
> I noticed that the temporary `go.mod` files that we generate are not cleaned up after `vendor.sh` completes. They are added to `.gitignore`, so not "critical", but perhaps (as a follow up?) we should clean those up, so that running `vendor.sh` leaves a clean state afterwards.

Maybe, but I am not quite sure why this matters. With these files written one can actually use `go mod` commands directly, without having to always use `vendor.sh`, which I find quite convenient.

thaJeztah commented:
> With these files written one can actually use go mod commands directly, without having to always use vendor.sh, which I find quite convenient.

Ah, good point as well, didn't think of that

thaJeztah commented:
One caveat with that is that it still requires the `-modfile` flag;

```
go mod graph | grep ' github.com/BurntSushi/toml'
# nothing

go mod graph -modfile=vendor.mod | grep ' github.com/BurntSushi/toml'
github.com/docker/docker github.com/BurntSushi/toml@v0.3.1
github.com/moby/buildkit@v0.8.2-0.20210615162540-9f254e18360a github.com/BurntSushi/toml@v0.3.1
github.com/urfave/cli@v1.22.2 github.com/BurntSushi/toml@v0.3.1
github.com/urfave/cli@v1.22.1 github.com/BurntSushi/toml@v0.3.1
github.com/urfave/cli@v1.22.4 github.com/BurntSushi/toml@v0.3.1
```

thaJeztah reviewed (approved):
LGTM

errordeveloper commented:
> There's also some dependencies that we previously were able to avoid (by only consuming certain packages), but are now pulled in because `go.mod` mentions them and/or because some repositories are using a `tools.go` (or equivalent) to have an easy way to download test dependencies.

I guess the way I think about this is that new sub-decencies are not a major concern, as long as these aren't actually imported, which I checked for this PR.

errordeveloper commented:
@thaJeztah thanks for sharing `go mod graph` results for new deps, I think you need to tweak formatting a little to work with `<details>`.

thaJeztah commented:
> I think you need to tweak formatting a little to work with `<details/>`.

🤦 I always forget the empty line; should be fixed now

tonistiigi commented on vendor.mod:
```diff
@@ -0,0 +1,208 @@
+	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
+	google.golang.org/api v0.46.0 // indirect
+	google.golang.org/appengine v1.6.7 // indirect
+	honnef.co/go/tools v0.0.0-20190523083050-ea95bdfd59fc // indirect
```
I see we are now vendoring some new packages that actually never get called. Nothing we can do about it?

thaJeztah replied:
No, don't think so unfortunately. Some of these may be test dependencies or a dependency using a "tools.go" file to force it being included. (also see my comment above)

tonistiigi reviewed (commented):
When I run `hack/validate/vendor` inside the container the `vendor.sum` file changes significantly. But the validation script itself returns with 0 exit code.

errordeveloper commented on hack/validate/vendor:
```diff
@@ -3,10 +3,15 @@
 	IFS=$'\n'
 	# shellcheck disable=SC2207
-	files=( $(validate_diff --diff-filter=ACMR --name-only -- 'vendor.conf' 'vendor/' || true) )
+	files=( $(validate_diff --diff-filter=ACMR --name-only -- 'vendor.sum' 'vendor.mod' 'vendor/' || true) )
```
@tonistiigi thanks for checking this, well spotted! I have looked into this and it doesn't seem to detect uncommitted changes to any of these files, it seems to be just getting a list of files that changed between `origin/HEAD` and `HEAD`.

When I ran `bash -x hack/validate/vendor`, the command it called was this:
```
+ git diff 088afc99e4bf8adb78e29733396182417d67ada2...0089b1ffb718e9c413a0687dd714dfc5c0829cb4 --diff-filter=ACMR --name-only -- vendor.sum vendor.mod vendor/
```
Where 088afc99e4bf8adb78e29733396182417d67ada2 is what's in master right now, and 0089b1ffb718e9c413a0687dd714dfc5c0829cb4 is the head of my branch.

tonistiigi replied:
Doesn't at least line 21 need `vendor.mod` and `vendor.sum` as well?

errordeveloper replied:
> Doesn't at least line 21 need `vendor.mod` and `vendor.sum` as well?

You are right, that looks like where detection of uncommitted changes happens.

I can also see what `validate_diff` is all about, it checks what files changed in `HEAD` vs `origin/HEAD`.
I just tend to use `git diff` to check for uncommitted changes, so I too quickly assumed that's what `validate_diff` might be for, the function name isn't very specific...

errordeveloper replied:
I've updated this script just now.

errordeveloper commented:
> When I run `hack/validate/vendor` inside the container the `vendor.sum` file changes significantly. But the validation script itself returns with 0 exit code.

I've committed an update to `vendor.sum` just now, see https://github.com/moby/moby/pull/43075#discussion_r773009628 regarding the behaviour of the validation script...

thaJeztah commented:
@errordeveloper I was a bit in doubt in which order we should merge the PR's; 

https://github.com/moby/moby/pull/42247 was green now (with the test-fixes merged), and has some commits that remove a couple of dependencies. I gave it a quick try to rebase this (your) PR on top of that one, to see how easy we can do a rebase (and update the `vendor.mod` etc accordingly; https://github.com/moby/moby/pull/43101

I can also do a try to see what the "reverse" will look like (have not tried yet)

errordeveloper commented:
@thaJeztah ok, do give it a go! I can pick this back up on Friday, if needed.

errordeveloper commented:
closing in favour of #43101
